### ISSUE TITLE
Whether/How to add new keywords post-2.0?

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#24801]
Following the 2.0 release, we may want to add some new reserved keywords to the language. Some examples of where this might come up:
* `abstract` to mark purely virtual methods
* `exscan` for exclusive scans
* `static` local variables
* the ongoing discussion about adding a new keyword to replace `dmapped`: https://github.com/chapel-lang/chapel/issues/23128

However, reserving a new word has the potential to break existing user code if the word is already used as an identifier. This issue asks whether we should be able to add new keywords post-2.0 and what the process should be in order to minimize the impact on users.

### Proposals:

1.  **warning approach:**
When a new keyword is needed to add/improve a language feature, we would implement a compiler warning that lasts for some number of release cycles (maybe 2). The warning would be emitted any time a code uses the word as an identifier, indicating that it will no longer be legal to do so in an upcoming release, and that the name of the variable/proc/enum/etc. should be changed.
After the warning has been in place for a number of releases, we would then introduce the new keyword and the associated feature. At this point, the parser would emit an error if the keyword is used in a syntactically-incorrect manner.

2. **opt-in approach:**
New keywords (and their associated features) would only be accessible under a new compiler flag (e.g., --chapel-3.0). This would allow users to keep using the stable set of chapel 2.0 features for the time being without the need to adjust their codes for breaking changes. Without throwing the flag, new features that rely on additional keywords would simply be unavailable. Throwing the flag would require renaming any identifiers that happen to share names with new keywords added since chapel 2.0.
We could also consider creating scripts or some other mechanism to help users migrate their codes to use the flag.
(this is similar to Rust's [editions approach](https://doc.rust-lang.org/edition-guide/editions/#what-are-editions) to adding new keywords)

---

For reference, [here](https://chapel-lang.org/docs/language/spec/lexical-structure.html#keywords) is the current list of reserved words.

**Comment by jabraham17** [comment]
I think this issue is resolved with editions. @lydia-duncan is that correct? We may need to do some more implementation for keyword stuff, but the design should be done

**Comment by bradcray** [comment]
I've been imagining (but am not confident) that new keywords could be added in an edition by having the lexer look for the keyword, and turn it into a token if the edition was selected or a normal TIDENTIFIER if it was not (in which case, assuming the edition flag is easy to check, there may not even need to be any special implementation to support it).  Does that sound right?

**Comment by lydia-duncan** [comment]
> I think this issue is resolved with editions. [@lydia-duncan](https://github.com/lydia-duncan) is that correct? We may need to do some more implementation for keyword stuff, but the design should be done

Yeah, I'd agree that the design is done and the result is "use editions to guard the new keyword".  I definitely didn't attempt to put support for parser changes in the initial implementation, so work would need to be done there (and that work is the more complicated of the compiler support for editions).  I think the list of keywords here is helpful so might continue to leave it open until that implementation occurs, but we can always close it and open a new issue that's focused on the implementation side.

> I've been imagining (but am not confident) that new keywords could be added in an edition by having the lexer look for the keyword, and turn it into a token if the edition was selected or a normal TIDENTIFIER if it was not (in which case, assuming the edition flag is easy to check, there may not even need to be any special implementation to support it). Does that sound right?

We hadn't done any discussion about the exact implementation details, to my memory.  My mental model had been more coarse-grained than that - I was thinking of having a different directory for parser files for each edition.  But that's more from the assumption that any parser changes would be breaking anyways.  If that's not the case, the code duplication would probably be too much of a maintenance burden.

The edition flag checking being done in the driver file was in anticipation of needing that information as early as possible.  So hopefully the flag is easy enough to check against at parse time, but we'll best know that when we actually do the implementation.

**Comment by bradcray** [comment]
> I was thinking of having a different directory for parser files for each edition.

I think we should avoid this at all costs.  The parser is complicated enough as it is, and I don't see any real advantage to introducing code clones of it if we can avoid them.

Just to put some code behind my previous comment, now that I've got a few more minutes, imagine:

```lex
// treat prototype as a normal identifier after the 2.0 edition
prototype        return (checkEditionOrEarlier("2.0")) ?
                   processToken(yyscanner, TPROTOTYPE); : 
                   processIdentifier(yyscanner, '\0');
```

where that pattern could potentially be put into some sort of helper like:

```
prototype        return identOrKeyword(CheckEditionOrEarlier("2.0"), TPROTOTYPE);
```

**Comment by lydia-duncan** [comment]
> The parser is complicated enough as it is, and I don't see any real advantage to introducing code clones of it if we can avoid them.

I think that's a matter of perspective.  If all the changes to it are breaking changes that need to be guarded by editions, then isn't any change that accounts for multiple editions within the same file adding more complexity than a simple check about which version to use?  Once the old editions we were supporting get dropped, we'll have to pull significantly more code out and we could miss some, leaving unnecessary cruft behind.  Meanwhile dropping support for an old edition with my proposal would leave much less room for old cruft getting littered in our code going forward as we would be able to delete a single file wholesale.  That feels much more like the K.I.S.S. design principle to me.

That said, I really don't feel strongly about this, especially since having a copy is a maintenance burden.  All it would take to convince me is one single, concrete example of a parser change that we would do in the future that isn't breaking for users.  I'm working from the perspective that the dyno parser improvements covered any cleanups we would have wanted to do that aren't breaking.  But we're human beings, so it's entirely possible there were things we didn't account for during that cleanup that will become apparent to someone examining it later.  And it's also possible that there were changes we wanted to do but couldn't due to needing to maintain compatibility with the production compiler.  An example along those lines would convince me.
